core/src/main/java/org/apache/druid/java/util/common/DefineClassUtils.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.common;
+
+import org.apache.druid.utils.JvmUtils;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.security.ProtectionDomain;
+
+/**
+ * This utility class provides a thin runtime abstraction to pick between
+ * - sun.misc.Unsafe.defineClass in Java 8,
+ * - and MethodHandles.Lookup.defineClass in Java 9 and above,
+ * while still providing compile-time support for both Java 8 and Java 9+.
+ *
+ * See also {@link ByteBufferUtils}
+ */
+public class DefineClassUtils
```

Comment:
Hi @xvrl, I'm trying to sort out what are missing in LICENSE and NOTICE files. Are this class and `UnsafeUtils` copied from Kafka?

Reply:
Or would you please let me know if there are any other classes copied from Kafka except ByteBufferUtils?

Reply:
Only portions of ByteBufferUtils are adapted from Apache Kafka. I already updated the LICENSE file in this PR https://github.com/apache/incubator-druid/pull/7466/files#diff-9879d6db96fd29134fc802214163b95aR269

I did not see any corresponding NOTICE changes for bits we adapted from other projects, so I followed the same model. cc @julianhyde you might be able to clarify the nuance of NOTICE/LICENSE.

Reply:
@xvrl thanks for reminding me the update in LICENSE file.

> If the Work includes a "NOTICE" text file as part of its distribution, then any Derivative Works that You distribute must include a readable copy of the attribution notices contained within such NOTICE file, excluding those notices that do not pertain to any part of the Derivative Works, in at least one of the following places: within a NOTICE text file distributed as part of the Derivative Works; within the Source form or documentation, if provided along with the Derivative Works; or, within a display generated by the Derivative Works, if and wherever such third-party notices normally appear. The contents of the NOTICE file are for informational purposes only and do not modify the License. You may add Your own attribution notices within Derivative Works that You distribute, alongside or as an addendum to the NOTICE text from the Work, provided that such additional attribution notices cannot be construed as modifying the License. 

From this statement (http://www.apache.org/licenses/LICENSE-2.0.html#redistribution), I think the notice for Kafka source distribution should be contained in Druid's NOTICE file. I'm working on that and will raise a PR soon.

Reply:
This part is important:

> excluding those notices that do not pertain to any part of the Derivative Works

It means that if the relevant Kafka code was originally written by a Kafka contributor (and not copied from somewhere else), then nothing from the Kafka NOTICE is needed.

Reply:
I agree, @gianm. Hopefully, and very likely, the code that was copied was developed as part of the regular Kafka development process (someone should check) and therefore does not have any NOTICE.

This, by the way, is why each project should keep its NOTICE as short as possible. To remove burden from downstream projects.

Reply:
Oh I meant we need to include only the below part instead of the whole notices:

```
================= Apache Kafka =================
Apache Kafka
Copyright 2019 The Apache Software Foundation
```

Reply:
the Apache Kafka code itself was inspired by Apache Lucene, as indicate in the commit message https://github.com/apache/kafka/commit/e554dc518eaaa0747899e708160275f95c4e525f so we may have to include a bit more info.
It looks like Kafka did not make a reference to this in their NOTICE or LICENSE file. cc @ijuma